Expressify the number types

thofma opened the pull request:
Fixes #322.

@tthsqe12: The polynomial printing is picking up our `show(io::IO, ::MIME{Symbol("text/plain")}, a::AbstractAlgebra.MPolyElem)` defined in AbstractAlgebra. So I think the best is to just define expressify for the Singular.jl number types. I did it already for `n_Z` and `n_Q` (which fixes #322).

I would also like to add it to `n_Zn`, `n_Zp` and `n_GF`. For the first two it would be the best if we could convert the vaules to `Int/BigInt`. Do you know if this is possible?

tthsqe12 commented:
Try to keep in mind singular's numbers are a lot like void pointers.

For z_Zn, it looks like it is just a mpz_ptr:
https://github.com/Singular/Singular/blob/spielwiese/libpolys/coeffs/rmodulon.cc#L160
and you can see some manual unpacking of the mpz struct here
https://github.com/oscar-system/Singular.jl/blob/master/src/number/n_Zn.jl#L32

For n_Zp, it looks like you just cast the pointer to a long (i.e. it is an immediate data type):
https://github.com/oscar-system/Singular.jl/blob/master/src/number/n_Zp.jl#L377

thofma commented:
Any clue what we should do about n_GF?

thofma commented:
OK, this is good to go from my part. I decided to not do `n_GF`.

@tthsqe12 Would be good if you could have a quick look, in particular the pointer fun.

fingolfin commented on src/number/n_Zn.jl:
```diff
@@ -89,6 +89,14 @@ function show(io::IO, c::N_ZnRing)
+function AbstractAlgebra.expressify(n::n_Zn; context = nothing)::Any
+  x = BigInt()
+  GC.@preserve n begin
+    ccall((:__gmpz_set, :libgmp), Cvoid, (Ref{BigInt},  Ptr{Any}), x, n.ptr.cpp_object)
```
This code is also not tested by the test suite, would be good if it was to check the `ccall`.

Also, I think it would be quite useful and natural to have a function that allows converting an `n_Zn` to `BigInt` (or `fmpz` or `n_Z`?). Perhaps we should copy/move the above code into a `BigInt` constructor that one can use to do so?

thofma replied:
I would rather not include the discussion on what `BigInt(::n_Zn)` returns in this PR here (lift to `[0, n)` or `[-n/2,n/2]`).

fingolfin commented on src/number/n_Z.jl:
```diff
@@ -102,6 +102,10 @@ function show(io::IO, c::Integers)
 end
 
+function AbstractAlgebra.expressify(n::n_Z; context = nothing)::Any
+  return AbstractAlgebra.expressify(BigInt(n), context = context)
```
Codecov complains that this code (and some of the other code you add) is not covered by tests. Could you perhaps add some basic tests, to verify these new methods work (and keep working)?

thofma commented:
I updated the PR. I also fixed it so that we have the same behavior as Singular:
```julia
julia> R = ResidueRing(ZZ, 5)
Residue Ring of Integer Ring modulo 5

julia> R(3)
3

julia> R = Fp(5)
Finite Field of Characteristic 5

julia> R(3)
-2
```

tthsqe12 commented:
I will look into n_GF at a later time. The symmetric remainder is good since that is what singular does.

tthsqe12 commented:
@thofma can you reproduce this bad behaviour?

```
julia> using Singular
Singular.jl, based on
                     SINGULAR                                 /
 A Computer Algebra System for Polynomial Computations       /  Singular.jl: 0.4.2-DEV
                                                           0<   Singular   : 4.2.0
 by: W. Decker, G.-M. Greuel, G. Pfister, H. Schoenemann     \
FB Mathematik der Universitaet, D-67653 Kaiserslautern        \
     

julia> R(,(x,y,z) = PolynomialRing(Fp(5),["x", "y", "z"])
ERROR: syntax: unexpected ","
Stacktrace:
 [1] top-level scope at none:1

julia> R,(x,y,z) = PolynomialRing(Fp(5),["x", "y", "z"])
(Singular Polynomial Ring (ZZ/5),(x,y,z),(dp(3),C), spoly{n_Zp}[x, y, z])

julia> (x+2*y+3*z)^2
1*x^2 + -1*x*y + -1*y^2 + 1*x*z + 2*y*z + -1*z^2

julia> Singular.AbstractAlgebra.expressify((x+2*y+3*z)^2)
:("1" * x ^ 2 + "-1" * x * y + "-1" * y ^ 2 + "1" * x * z + "2" * y * z + "-1" * z ^ 2)

julia> Singular.AbstractAlgebra.expressify(Fp(5)(3))
"-2"

julia> Fp(5)(3)
-2

julia> @which Singular.AbstractAlgebra.expressify(Fp(5)(3))
expressify(a; context) in AbstractAlgebra at /home/schultz/.julia/packages/AbstractAlgebra/aF5Iw/src/PrettyPrinting.jl:57
```

tthsqe12 reviewed (approved):
looks good - I will add n_GF shortly
